framework/src/araui/assets/ui_font_asset.c

```diff
@@ -0,0 +1,280 @@
+		return UI_NULL;
+	}
+
+	// file open
+	file = fopen(filename, "r");
+	if (!file) {
+		UI_LOGE("error: failed to open file!\n");
+		return UI_NULL;
+	}
+
+	// read header of file
+	ui_fread(&font_height, 4, 1, file);
+	ui_fread(&glyph_num, 4, 1, file);
+
+	// make font asset body
+	body = (ui_font_asset_body_t *)UI_ALLOC(sizeof(ui_font_asset_body_t));
+	if (!body) {
+		(void)fclose(file);
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_font_asset_body_t));
+	((ui_asset_body_t *)body)->type = UI_FONT_ASSET;
+	body->height = font_height;
+
+	// read glyph from the buffer
+	ui_map_init(&body->glyph_map, DEFAULT_GLYPH_MAP_CAPACITY);
+	for (glyph_idx = 0; glyph_idx < glyph_num; glyph_idx++) {
+		ui_fread(&utf8_code, 4, 1, file);
+		ui_fread(&font_width, 4, 1, file);
+		ui_fread(&data_size, 4, 1, file);
+
+		glyph = (ui_asset_glyph_t *)UI_ALLOC(sizeof(ui_asset_glyph_t));
+		if (!glyph) {
+			_ui_font_asset_destroy_func((void *)body);
+			(void)fclose(file);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		glyph->data = (uint8_t *)UI_ALLOC(data_size);
+		if (!glyph->data) {
+			UI_FREE(glyph);
+			_ui_font_asset_destroy_func((void *)body);
+			(void)fclose(file);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		ui_fread(glyph->data, 1, data_size, file);
+
+		glyph->size = data_size;
+		glyph->width = font_width;
+
+		ui_map_push(&body->glyph_map, utf8_code, glyph);
+	}
+
+	(void)fclose(file);
+
+	return (ui_asset_t)body;
+}
+
+ui_asset_t ui_font_asset_create_from_buffer(const uint8_t *buf)
+{
+	ui_font_asset_body_t *body;
+	ui_font_data_t *font_data;
+	ui_font_glyph_data_t *glyph_data;
+	uint32_t glyph_num;
+	int glyph_idx;
+	ui_asset_glyph_t *glyph;
+
+	if (!ui_is_running()) {
+		UI_LOGE("error: UI framework is not running!\n");
+		return UI_NULL;
+	}
+
+	if (!buf) {
+		UI_LOGE("error: invalid parameter!\n");
+		return UI_NULL;
+	}
+
+	body = (ui_font_asset_body_t *)UI_ALLOC(sizeof(ui_font_asset_body_t));
+	if (!body) {
+		UI_LOGE("error: out of memory!\n");
+		return UI_NULL;
+	}
+
+	memset(body, 0, sizeof(ui_font_asset_body_t));
+	((ui_asset_body_t *)body)->type = UI_FONT_ASSET;
+
+	font_data = (ui_font_data_t *)buf;
+	glyph_num = font_data->glyph_count;
+	body->height = font_data->font_height;
+	body->from_buf = true;
+
+	glyph_data = (ui_font_glyph_data_t *)(buf + font_data->header_size);
+
+	// read glyph from the buffer
+	ui_map_init(&body->glyph_map, DEFAULT_GLYPH_MAP_CAPACITY);
+	for (glyph_idx = 0; glyph_idx < glyph_num; glyph_idx++) {
+		glyph = (ui_asset_glyph_t *)UI_ALLOC(sizeof(ui_asset_glyph_t));
+		if (!glyph) {
+			_ui_font_asset_destroy_func((void *)body);
+			UI_LOGE("error: out of memory!\n");
+			return UI_NULL;
+		}
+
+		glyph->data = (uint8_t *)glyph_data + glyph_data->header_size;
+		glyph->size = glyph_data->data_size;
+		glyph->width = glyph_data->font_width;
+
+		ui_map_push(&body->glyph_map, glyph_data->utf8_code, glyph);
```

Comment:
What if this function fails?
Should be considered.

Reply:
I'll add return type and the error logs.
This logic can skip the error (only the logging is needed).
Because unadded glyph is not showing.